[API] Unvirtualize layer setDataSource method

rldhont opened the pull request:
## Description

Changes requested in #41208  in the layer API:
* The setDatasource() methods become non-virtual public method
* The virtual private setDataSourcePrivate() method is defined and the provider setDataSource implementations have been moved to this new virtual method

The first step is [Tests] Fix deprecated QgsVectorLayer::setDataSource usage

The second one is [API] Remove specific deprecated QgsVectorLayer::setDataSource

rldhont committed:
[API] Unvirtualize layer setDataSource method

Changes in the layer API:
* The setDatasource() methods become non-virtual public method
* The virtual private setDataSourcePrivate() method is defined and the provider setDataSource implementations have been moved to this new virtual method

rldhont committed:
[API] Remove specific deprecated QgsVectorLayer::setDataSource

rldhont committed:
[Tests] Fix deprecated QgsVectorLayer::setDataSource usage

m-kuhn commented:
> The setDatasource() methods become non-virtual public method

For reference.
I think we cannot remove virtual from the existing one until qgis4 it would be an api break, but we should not add another one and get all usage in `our` own codebase right to be prepared.

nyalldawson commented:
@m-kuhn 

Do we really want plugins overriding this? I can't see any situation where it's a good thing and wouldn't result in a ton of breakage

m-kuhn commented:
Not at all. I'm in favor of getting rid of it asap. Api stability?

nyalldawson commented:
We've definitely had cases in the past where we've technically broken stable api by changing methods which were never intended to be used by third parties and where doing so is dangerous. I'd say this qualifies.

m-kuhn commented:
Works for me !

m-kuhn commented:
@rldhont in this case we need to list it in the api break file, but I'd be as explicit as possible to avoid misinterpretation:

> QgsMapLayer::setDataSource is no longer virtual. If has been replaced with QgsMapLayer::setDataSourcePrivate. This is only relevant for subclassing QgsMapLayer, for code that only uses the subclasses (QgsVectorLayer, QgsRasterLayer, ...) this has no effect.

rldhont committed:
[Tests] TestPyQgsQgsQueryResultModel - split assertTrue into assertGreater and assertLess

The test_model_stop in TestPyQgsQgsQueryResultModel can failed even if nothing has changed in the code.
The `(res.fetchedRowCount() > 0 and res.fetchedRowCount() < self.NUM_RECORDS)` can return False instead of True and the error message is not enough specific.
To get more specific error message the asserTrue has been split into assertGreater and assertLess.

Gustry commented:
@rldhont https://github.com/qgis/QGIS/blob/master/doc/api_break.dox

rldhont commented:
Thanks @Gustry

rldhont commented:
Thanks @nyalldawson @m-kuhn, It is ready for review.

m-kuhn commented on tests/src/python/test_qgsprojectbadlayers.py:
```diff
@@ -219,8 +219,9 @@ def _check_relations():
-        point_a.setDataSource(point_a_source, 'point_a', 'ogr')
-        point_b.setDataSource(point_b_source, 'point_b', 'ogr')
+        options = QgsDataProvider.ProviderOptions()
+        point_a.setDataSource(point_a_source, 'point_a', 'ogr', options)
```
Can we add this as a new test?
The old way should continue to work.

rldhont replied:
The old way does not works, it is based on an unvirtual deprecated method only defined for QgsVectorLayer 88c0d3f07db541b1cb6d5cc81f4aabc12741de26

I can move this method to QgsMapLayer to keep API

m-kuhn replied:
Can't we keep it where it was?

rldhont replied:
If I keep it where it was, the compilation failed because this method is marked as deprecated.

m-kuhn replied:
This can normally be avoided by `Q_NOWARN_DEPRECATED_PUSH` / `Q_NOWARN_DEPRECATED_POP`
Removing it is an API break, so we can't do that.
Moving it to QgsMapLayer is not nice because it increases exposure of a deprecated API, ideally we should avoid that.

rldhont replied:
It seems that keeping the dedicated QgsVectorLayer::setDataSource method in place. The compilation failed.

I need to explicitly call the QgsMapLayer::setDataSource method

rldhont replied:
It seems that keeping the dedicated QgsVectorLayer::setDataSource method in place. It is not possible to use QgsMapLayer::setDataSource in PyQGIS from QgsVectorlayer instance.

m-kuhn commented on tests/src/python/test_provider_shapefile.py:
```diff
@@ -629,7 +630,7 @@ def testOpenWithFilter(self):
             self.assertEqual(f['name'], 'Apple')
             # force close of data provider
-            vl.setDataSource('', 'test', 'ogr')
+            vl.setDataSource('', 'test', 'ogr', options)
```
Can we add this as a new test to be sure the old way still works

rldhont replied:
The old way does not works, it is based on an unvirtual deprecated method only defined for QgsVectorLayer 88c0d3f07db541b1cb6d5cc81f4aabc12741de26

I can move this method to QgsMapLayer to keep API

rldhont replied:
The old way is a deprecated way, do we need to keep test that use deprecated method ?

m-kuhn commented on src/core/vector/qgsvectorlayer.h:
```diff
@@ -2856,6 +2830,19 @@ class CORE_EXPORT QgsVectorLayer : public QgsMapLayer, public QgsExpressionConte
+     * \param baseName base name of the layer
+     * \param provider provider string
+     * \param options provider options
+     * \param flags provider read flags
```
Would you be able to improve the documentation of the parameters here? I see they are copied, so if it's out of reach just say so.

rldhont replied:
Done

rldhont committed:
Revert "[API] Remove specific deprecated QgsVectorLayer::setDataSource"

This reverts commit 88c0d3f07db541b1cb6d5cc81f4aabc12741de26.

rldhont committed:
Update method doc

rldhont committed:
Revert "Revert "[API] Remove specific deprecated QgsVectorLayer::setDataSource""

This reverts commit 2524f9635a3d09c57ccebb78208cc7b3bf8fb9b6.

rldhont committed:
[API] Move specific deprecated QgsVectorLayer::setDataSource to QgsMapLayer

rldhont committed:
fix move

rldhont commented:
@m-kuhn I have move the specific old QgsVectorLayer::setDataSource marked as deprecated since 3.2 to QgsMapLayer to avoid API break in QgsVectorLayer.

m-kuhn commented on python/core/auto_generated/vector/qgsvectorlayer.sip.in:
```diff
@@ -1082,40 +1082,6 @@ Determines if this vector layer has features.
 .. versionadded:: 3.4
 %End
 
- void setDataSource( const QString &dataSource, const QString &baseName, const QString &provider, bool loadDefaultStyleFlag = false ) /Deprecated/;
```
This one is also an API break, no?

rldhont replied:
It is not an API break because the method is provide by QgsMapLayer, so the method for QgsVectorLayer instances is always available.

If I want to unvirtualize QgsMapServer::setDataSource methods, I have to remove the specific QgsVectorLayer::setDataSource. And to avoid an other API break than unvirtualize, I move the method to QgsMapLayer.

m-kuhn reviewed (approved):
Looks all good to me, no more (unacceptable) api breaks and a very nice stabilization of the api on the other hand.
Thanks !

m-kuhn committed:
Merge branch 'master' into unvirtualize-setdatasource

rldhont commented:
Thanks @m-kuhn for the review and merge

elpaso commented on src/core/vector/qgsvectorlayer.cpp:
```diff
@@ -183,7 +183,12 @@ QgsVectorLayer::QgsVectorLayer( const QString &vectorLayerPath,
+    {
+      providerFlags |= QgsDataProvider::FlagLoadDefaultStyle;
+    }
+    setDataSourcePrivate( vectorLayerPath, baseName, providerKey, providerOptions, providerFlags );
```
@m-kuhn my compiler throws a warning here:

![immagine](https://user-images.githubusercontent.com/142164/118009201-2684f200-b34e-11eb-8b82-fff75b4ec081.png)

rldhont replied:
Is the commit https://github.com/qgis/QGIS/pull/41208/commits/dd2edf0bf15f2017f36cc2ac12990261249a0d92 in https://github.com/qgis/QGIS/pull/41208 fixed the warning ? I have replaced `setDataSourcePrivate` call by `setDataSource` ?
